### PR TITLE
HPCC-8710 Multi slave in same process framework

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith jake.smith@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
NB: based on 2 as yet unmerged branches (hence 3 commits)

Still testing on a multi node system, but passes regression suite on 1 node with multiple slaves.

@richardkchapman @ghalliday - please review.

File: system/mp/mpcomm.cpp [hunk 1]
```diff
@@ -2976,8 +3030,6 @@ void stopMPServer()
         MPserver->stop();
         delete MPserver;
         MPserver = NULL;
-        ::Release(worldcomm);
```
Comment (ghalliday) [inline system/mp/mpcomm.cpp]: This will now never be cleaned up.

Comment (jakesmith) [inline system/mp/mpcomm.cpp]: Addressed by separate PR (gh-7812).
However, not sure why I took these lines out, still better to clear at this point. I will reinstate (in PR for HPCC-14244)

File: thorlcr/graph/thgraphmaster.cpp [hunk 2]
```diff
@@ -153,11 +154,14 @@ void CSlaveMessageHandler::main()
                             graph->fireException(e);
                     }
                     else
+                    {
+                        e->setSlave(slave);
```
Comment (ghalliday) [inline thorlcr/graph/thgraphmaster.cpp]: This has already been done above.

Comment (jakesmith) [inline thorlcr/graph/thgraphmaster.cpp]: true, not new, but I'll remove that line.

File: thorlcr/graph/thgraphmaster.cpp [hunk 3]
```diff
@@ -2519,44 +2566,48 @@ void CMasterGraph::getFinalProgress()
         if (0 == msg.remaining())
             continue;
 
-        handleSlaveDone(sender-1, msg);
-
-        if (!queryOwner())
+        unsigned slavesPerProcess = globals->getPropInt("@slavesPerProcess", 1); // JCSMORE - should move somewhere common
```
Comment (ghalliday) [inline thorlcr/graph/thgraphmaster.cpp]: I think this is testing the wrong variable.   Should it be some combination of slavesPerNode and processPerSlave?

**Comment by ghalliday** [comment]
Minor: It might help clarity some of the code if a new slave_t type was introduced.  Sometimes it is hard to tell from the function if it is a slave index, or an index within the group.

File: thorlcr/master/thgraphmanager.cpp [hunk 4]
```diff
@@ -288,6 +288,14 @@ void CJobManager::run()
         dp->qn = queueName.get();
     }
 
+    PROGLOG("verifying mp connection to all slaves");
```
Comment (ghalliday) [inline thorlcr/master/thgraphmanager.cpp]: Does this impose any significant overhead?

File: thorlcr/graph/thgraphmaster.cpp (hunk 3 above)
Comment (jakesmith) [inline thorlcr/graph/thgraphmaster.cpp]: Yes, there was a point where it existed, but had changed already.

File: thorlcr/master/thgraphmanager.cpp (hunk 4 above)
Comment (jakesmith) [inline thorlcr/master/thgraphmanager.cpp]: The idea is that it helps, by  doing upfront rather then when needed.
This is a one-time thing (not per job), making all connection to each other.

File: initfiles/componentfiles/configxml/thor.xsd.in [hunk 5]
```diff
@@ -483,21 +483,21 @@
...
+      <xs:attribute name="slavesPerNode" type="xs:nonNegativeInteger" use="optional" default="1">
         <xs:annotation>
           <xs:appinfo>
-            <tooltip>Base port for local thor slaves (when localThor set)</tooltip>
+            <tooltip>This allows multiple slaves to exist on each node</tooltip>
           </xs:appinfo>
         </xs:annotation>
       </xs:attribute>
-      <xs:attribute name="slavesPerNode" type="xs:nonNegativeInteger" use="optional" default="1">
+      <xs:attribute name="processPerSlave" type="xs:xs:boolean" use="optional" default="false">
```
Comment (ghalliday) [inline initfiles/componentfiles/configxml/thor.xsd.in]: Should this default to true for backward compatibility?

File: thorlcr/slave/slavmain.cpp [hunk 6]
```diff
@@ -153,21 +157,70 @@ class CJobListener : public CSimpleInterface
...
     CJobListener() : excptHandler(*this)
     {
         stopped = true;
+        processPerSlave = globals->getPropBool("@processPerSlave");
+        slavesPerNode = globals->getPropInt("@slavesPerNode", 1);
+        mpServers.append(* getMPServer());
+        if (!processPerSlave)
+        {
+            unsigned port = getMachinePortBase();
+            // will need generating into thor.xml, currently only in setvars
```
Comment (ghalliday) [inline thorlcr/slave/slavmain.cpp]: Is this comment still valid?

File: thorlcr/slave/slavmain.cpp [hunk 7]
```diff
@@ -153,21 +157,70 @@ class CJobListener : public CSimpleInterface
...
+                }
+                ~CVerifyThread() { join(); }
+                void start() { threaded.start(); }
+                void join() { threaded.join(); }
+                virtual void main()
+                {
+                    Owned<ICommunicator> comm = jobListener.mpServers.item(channel).createCommunicator(&queryClusterGroup());
+                    PROGLOG("verifying mp connection to rest of slaves (from channel=%d)", channel);
+                    if (!comm->verifyAll())
+                        ERRLOG("Failed to connect to reset of slaves");
```
Comment (ghalliday) [inline thorlcr/slave/slavmain.cpp]: typo: reset->rest

File: thorlcr/slave/slavmain.cpp [hunk 8]
```diff
@@ -153,21 +157,70 @@ class CJobListener : public CSimpleInterface
...
+                    else
+                        PROGLOG("verified mp connection to rest of slaves");
+                }
+            };
+            CIArrayOf<CInterface> verifyThreads;
+            for (unsigned c=0; c<slavesPerNode; c++)
```
Comment (ghalliday) [inline thorlcr/slave/slavmain.cpp]: Should this block of code be executed on slave 0 if processPerSlave is false, or should this loop start from 1?

File: thorlcr/slave/slavmain.cpp [hunk 9]
```diff
@@ -651,6 +731,10 @@ class CThorResourceSlave : public CThorResourceBase
     virtual IFileInProgressHandler &queryFileInProgressHandler() { return *fipHandler.get(); }
 };
 
+void initSlaveMPServers()
```
Comment (ghalliday) [inline thorlcr/slave/slavmain.cpp]: I think this is unused.  Can probably be deleted.

File: thorlcr/slave/slwatchdog.cpp [hunk 10]
```diff
@@ -91,7 +91,7 @@ class CGraphProgressHandlerBase : public CSimpleInterface, implements ISlaveWatc
         {
             MemoryBuffer progressData;
             { CriticalBlock b(crit);
```
Comment (ghalliday) [inline thorlcr/slave/slwatchdog.cpp]: unrelated: indentation isn't consistent

File: thorlcr/slave/thslavemain.cpp [hunk 11]
```diff
@@ -85,7 +85,7 @@ static void replyError(unsigned errorCode, const char *errorMsg)
     Owned<IException> e = MakeStringException(errorCode, "%s", str.str());
     CMessageBuffer msg;
     serializeException(e, msg);
-    queryClusterComm().send(msg, 0, MPTAG_THORREGISTRATION);
+    queryWorldCommunicator().send(msg, 0, MPTAG_THORREGISTRATION);
```
Comment (ghalliday) [inline thorlcr/slave/thslavemain.cpp]: Why not queryNodeCommunicator() - is there a difference?

File: initfiles/componentfiles/configxml/thor.xsd.in [hunk 12]
```diff
@@ -462,14 +462,14 @@
...
-      <xs:attribute name="masterport" type="xs:nonNegativeInteger" default="20000">
+      <xs:attribute name="masterport" type="xs:nonNegativeInteger" default="20000" use="optional">
         <xs:annotation>
           <xs:appinfo>
             <tooltip>Base port to use for master</tooltip>
           </xs:appinfo>
         </xs:annotation>
       </xs:attribute>
-      <xs:attribute name="slaveport" type="xs:nonNegativeInteger" use="optional">
+      <xs:attribute name="slaveport" type="xs:nonNegativeInteger" default="20100" use="optional" >
```
Comment (ghalliday) [inline initfiles/componentfiles/configxml/thor.xsd.in]: trivial: space on end

Comment (jakesmith) [inline initfiles/componentfiles/configxml/thor.xsd.in]: will remove

File: initfiles/componentfiles/configxml/thor.xsd.in (hunk 5 above)
Comment (jakesmith) [inline initfiles/componentfiles/configxml/thor.xsd.in]: probably, I'll change.

File: thorlcr/slave/slavmain.cpp (hunk 6 above)
Comment (jakesmith) [inline thorlcr/slave/slavmain.cpp]: No, will remove.

File: thorlcr/slave/slavmain.cpp (hunk 7 above)
Comment (jakesmith) [inline thorlcr/slave/slavmain.cpp]: corrected

File: thorlcr/slave/slavmain.cpp (hunk 8 above)
Comment (jakesmith) [inline thorlcr/slave/slavmain.cpp]: No, because it has already done a verifyAll on the nodeCommunicator by this stage.
slave 0 above share the same IMPServer as the nodeCommunicator on each node, so will already be connected.

File: thorlcr/slave/slavmain.cpp (hunk 9 above)
Comment (jakesmith) [inline thorlcr/slave/slavmain.cpp]: true, will remove.

File: thorlcr/slave/slavmain.hpp [hunk 13]
```diff
@@ -19,6 +19,7 @@
 #define SLAVMAIN_HPP
 
 void abortSlave();
+void initSlaveMPServers();
```
Comment (ghalliday) [inline thorlcr/slave/slavmain.hpp]: can be deleted along with body above

File: thorlcr/slave/thslavemain.cpp (hunk 11 above)
Comment (jakesmith) [inline thorlcr/slave/thslavemain.cpp]: the world comm, can receive from anyone, the nodecomm can receive from only the thorgroup..
So could be nodecomm().

The idea originally was others could dynamically register, so needed to list to 'world'.
I changed it to queryWorldCommunicator() because it matched what slave was using, which doesn't know the group yet..

**Comment by ghalliday** [comment]
@jakesmith the changes look remarkably clean.  I have a couple more things to check, but looks close to ready.

**Comment by jakesmith** [comment]
@ghalliday I am investigating a couple of mismatches, it is not clear yet whether actually related or not though.

**Comment by mckellyln** [comment]
Yes, this is fantastic.

File: thorlcr/graph/thgraphslave.cpp [hunk 14]
```diff
@@ -965,10 +969,10 @@ class CThorCodeContextSlave : public CThorCodeContextBase, implements IEngineCon
         e->setOrigin(source);
         e->setAction(tea_warning);
         e->setSeverity((ErrorSeverity)severity);
-        job.fireException(e);
+        jobChannel.fireException(e);
     }
-    virtual unsigned getNodes() { return job.queryJobGroup().ordinality()-1; }
-    virtual unsigned getNodeNum() { return job.queryMyRank()-1; }
+    virtual unsigned getNodes() { return jobChannel.queryJob().queryJobGroup().ordinality()-1; }
```
Comment (ghalliday) [inline thorlcr/graph/thgraphslave.cpp]: observation:  could use queryJob().querySlaves();  It would have been slightly easier to check when reviewing.

**Comment by ghalliday** [comment]
@jakesmith one final question....
Are the calls to createReplyTag() in mpcomm.cpp correct, or do they need to know about the associated MpServer?

File: thorlcr/graph/thgraphslave.cpp (hunk 14 above)
Comment (jakesmith) [inline thorlcr/graph/thgraphslave.cpp]: It could/should, I'll change.

**Comment by jakesmith** [comment]
>Are the calls to createReplyTag() in mpcomm.cpp correct, or do they need to know about the associated MpServer?

@ghalliday - They should all be calling the CMPServer::createReplyTag() version.

**Comment by jakesmith** [comment]
@ghalliday - added a commit to make minor improvement to getNodes()

Update on regression mismatches, none were related to this commit.

File: initfiles/componentfiles/configxml/thor.xsd.in [hunk 15]
```diff
@@ -462,14 +462,14 @@
           </xs:appinfo>
         </xs:annotation>
       </xs:attribute>
-      <xs:attribute name="masterport" type="xs:nonNegativeInteger" default="20000">
+      <xs:attribute name="masterport" type="xs:nonNegativeInteger" default="20000" use="optional">
```
Comment (mckellyln) [inline initfiles/componentfiles/configxml/thor.xsd.in]: I thought there was a recent unrelated change to make this not optional so that it was always present in the xml for nagios and other external programs to be able to determine the thormaster port to scan.

Comment (jakesmith) [inline initfiles/componentfiles/configxml/thor.xsd.in]: I am not entirely sure why I changes this.. possibly unintentionally.
Will check and probably revert.

**Comment by ghalliday** [comment]
@jakesmith the file I should have referenced was hrpcmp.cpp

MpInterClientState and MpInterServerState

**Comment by jakesmith** [comment]
@ghalliday , do you see MpInterClientState and MpInterServerState used somewhere?
But it's 'ok' that they use the global MPserver createReply tags, as not used by any thing in Thor (or anything anywhere afaics).

They should be deleted (but don't need to be part of this PR)

**Comment by ghalliday** [comment]
I think you're right - it looks like that whole file could be deleted.
typo in commit message, but I suspect you'll squash.
